File: app/src/main/java/com/travlog/android/apps/viewmodels/EditNoteViewModel.java
package com.travlog.android.apps.viewmodels;

import android.annotation.SuppressLint;
import android.support.annotation.NonNull;

import com.travlog.android.apps.libs.ActivityViewModel;
import com.travlog.android.apps.libs.Environment;
import com.travlog.android.apps.libs.rx.Optional;
import com.travlog.android.apps.libs.rx.bus.NoteEvent;
import com.travlog.android.apps.models.Note;
import com.travlog.android.apps.services.ApiClientType;
import com.travlog.android.apps.ui.activities.EditNoteActivity;
import com.travlog.android.apps.viewmodels.errors.EditNoteViewModelErrors;
import com.travlog.android.apps.viewmodels.inputs.EditNoteViewModelInputs;
import com.travlog.android.apps.viewmodels.outputs.EditNoteViewModelOutputs;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.CompletableSubject;
import io.reactivex.subjects.PublishSubject;

import static com.travlog.android.apps.libs.rx.transformers.Transformers.neverApiError;
import static com.travlog.android.apps.libs.rx.transformers.Transformers.neverError;
import static com.travlog.android.apps.ui.IntentKey.NOTE;

public class EditNoteViewModel extends ActivityViewModel<EditNoteActivity>
        implements EditNoteViewModelInputs, EditNoteViewModelOutputs, EditNoteViewModelErrors {

    private final ApiClientType apiClient;

    private Note note;

    @SuppressLint("CheckResult")
    public EditNoteViewModel(final @NonNull Environment environment) {
        super(environment);

        this.apiClient = environment.apiClient;

        intent()
                .filter(i -> i.getParcelableExtra(NOTE) != null)
                .map(i -> (Note) i.getParcelableExtra(NOTE))
                .doOnNext(note -> this.note = note)
                .subscribe(note -> {
                    setTitleText.onNext(note.title);
                });

        title
                .compose(bindToLifecycle())
                .subscribe(title -> note.title = title);

        saveClick
                .switchMap(__ -> this.update(this.note)
                        .doOnSubscribe(disposable -> {

                        })
                        .doAfterTerminate(() -> {

                        }))
                .subscribe(this::updateSuccess);
    }

    private void updateSuccess(final @NonNull Note note) {
        NoteEvent.getInstance().post(note);
        back.onComplete();
    }

    private @NonNull
    Observable<Note> update(final @NonNull Note note) {
        return apiClient.updateNote(note)
                .compose(neverApiError())
                .compose(neverError())
                .toObservable();
    }

    private final PublishSubject<String> title = PublishSubject.create();
    private final PublishSubject<Optional> saveClick = PublishSubject.create();

    private final BehaviorSubject<String> setTitleText = BehaviorSubject.create();
    private final CompletableSubject back = CompletableSubject.create();

    public final EditNoteViewModelInputs inputs = this;
    public final EditNoteViewModelOutputs outputs = this;
    public final EditNoteViewModelErrors errors = this;

    @Override
    public void title(@NonNull String title) {
        this.title.onNext(title);
    }

    @Override
    public void saveClick() {
        this.saveClick.onNext(new Optional<>(null));
    }

    @NonNull
    @Override
    public Observable<String> setTitleText() {
        return setTitleText;
    }

    @NonNull
    @Override
    public Completable back() {
        return back;
    }
}

File: app/src/main/java/com/travlog/android/apps/viewmodels/EditNoteViewModel.kt
/*
 * Copyright 2019 Travlog. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.travlog.android.apps.viewmodels

import android.annotation.SuppressLint
import com.travlog.android.apps.libs.ActivityViewModel
import com.travlog.android.apps.libs.Environment
import com.travlog.android.apps.libs.db.realm.RealmHelper
import com.travlog.android.apps.libs.rx.Optional
import com.travlog.android.apps.libs.rx.bus.NoteEvent
import com.travlog.android.apps.libs.rx.transformers.Transformers.neverApiError
import com.travlog.android.apps.libs.rx.transformers.Transformers.neverError
import com.travlog.android.apps.models.Note
import com.travlog.android.apps.services.ApiClientType
import com.travlog.android.apps.ui.IntentKey.NOTE_ID
import com.travlog.android.apps.ui.activities.EditNoteActivity
import com.travlog.android.apps.viewmodels.errors.EditNoteViewModelErrors
import com.travlog.android.apps.viewmodels.inputs.EditNoteViewModelInputs
import com.travlog.android.apps.viewmodels.outputs.EditNoteViewModelOutputs
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.CompletableSubject
import io.reactivex.subjects.PublishSubject
import javax.inject.Inject

@SuppressLint("CheckResult")
class EditNoteViewModel @Inject constructor(environment: Environment
) : ActivityViewModel<EditNoteActivity>(environment),
        EditNoteViewModelInputs, EditNoteViewModelOutputs, EditNoteViewModelErrors {

    private val apiClient: ApiClientType = environment.apiClient

    private var note: Note? = null

    private val title = PublishSubject.create<String>()
    private val saveClick = PublishSubject.create<Optional<*>>()

    private val setTitleText = BehaviorSubject.create<String>()
    private val back = CompletableSubject.create()

    val inputs: EditNoteViewModelInputs = this
    val outputs: EditNoteViewModelOutputs = this
    val errors: EditNoteViewModelErrors = this

    init {
        intent()
                .compose(bindToLifecycle())
                .map { i -> i.getStringExtra(NOTE_ID) ?: "" }
                .map { RealmHelper.getNote(realm, it) }
                .doOnNext { this.note = it }
                .map { it.title }
                .subscribe { setTitleText.onNext(it) }

        title
                .compose(bindToLifecycle())
                .subscribe { note!!.title = it }

        saveClick
                .switchMap {
                    this.update(this.note!!)
                            .doOnSubscribe {

                            }
                            .doAfterTerminate {

                            }
                }
                .subscribe(this::updateSuccess)
    }

    private fun updateSuccess(note: Note) {
        NoteEvent.getInstance().post(note)
        back.onComplete()
    }

    private fun update(note: Note): Observable<Note> {
        return apiClient.updateNote(note)
                .compose(neverApiError())
                .compose(neverError())
                .toObservable()
    }

    override fun title(title: String) {
        this.title.onNext(title)
    }

    override fun saveClick() {
        this.saveClick.onNext(Optional<Any>(null))
    }

    override fun setTitleText(): Observable<String> {
        return setTitleText
    }

    override fun back(): Completable {
        return back
    }
}
